website/src/components/ExpoSnackLandingPage.js

```diff
@@ -17,14 +17,28 @@ import RedirectStarterSnack from './RedirectStarterSnack';
 
 export default function ExpoSnackLandingPage({match}) {
   const {expoSnackPath} = match.params;
+  const title = React.useMemo(
```

Comment:
This can't work at build time. 
`match.params` is probably an empty object there because Docusaurus does not know all the paths it has to build


view-source:https://playtorch-git-fork-raedle-export-d38482624-fbopensource.vercel.app/snack/@raedle/image-classification/index.html

Note: I don't understand why you don't get something like `undefined | PlayTorch Snack` 😅 that's worth opening the static HTML file locally and see what it contains.

---



But this works at runtime, and Docusaurus is able to "fix" your wrong metadata after React is loaded and hydrated: https://playtorch-git-fork-raedle-export-d38482624-fbopensource.vercel.app/snack/@raedle/image-classification/

![CleanShot 2022-08-10 at 11 41 20@2x](https://user-images.githubusercontent.com/749374/183869958-92b31838-a702-4133-a7e5-89754517f1df.png)


---

Unfortunately, I believe social previews require correct metadata to be set correctly at build time. They probably try to optimize their crawlers and only load the HTML file, not executing any JS.

So you really need to produce one HTML file per stack if you want social previews to work. I don't think this is compatible with dynamic routes.


You can just do something like:

```js
staticSnackList.forEach(snack => addRoute({
  path: "/snack/" + snack.path,
  exact: true,
}})

Reply:
Thanks @slorber for taking the time looking into this!

The Snacks aren't static and not known in advance, which is why the following won't work in our case unfortunately :/

```javascript
staticSnackList.forEach(snack => addRoute({
  path: "/snack/" + snack.path,
  exact: true,
}});
```

The dynamic `snack` route is a landing page for PlayTorch users to share their Snacks with others in the community. For example, if user A creates a Snack (e.g., `@userA/foobar`), the dynamic route is https://playtorch.dev/snack/@userA/foobar (foobar is the Snack named by the user). However, we won't know all the Snacks and Snack names that will be created by PlayTorch users.

As you mentioned and looking at the docs [Understanding SSR](https://docusaurus.io/docs/advanced/ssg#understanding-ssr), it seems like we need to break out of Docusaurus and implement a custom service that returns SSR pages with the correct Open Graph metadata.

Thanks again for your response!

Reply:
Yes unfortunately it looks like you can't use static site generator for this use-case.

Note that Vercel and others (Cloudflare, Netlify) have edge runtimes now, which means you can also write some code to "customize" the static HTML being served to the user. Even if the file has generic metadata, technically you could use request info to update the response and serve a different response per request.

I don't know if Vercel has that already (@leerob probably knows) but I've definitively seen this used on Cloudflare Workers: https://developers.cloudflare.com/workers/examples/modify-response/